### DISCUSSION TITLE
How do I return a Stream from a server function?

### DISCUSSION THREAD

**Discussion by eliaxelang007** [leptos-rs/leptos#3210] [Q&A] [answered]
Hello Leptos community!
I'm a bit new here, but I've been loving Leptos experience so far. :D

I have a question though: How do I return a Stream from a server function?
I know that it's possible because of these:

[Allow for a streaming value with #server.](https://github.com/leptos-rs/leptos/issues/1284)
[Leptos v0.6 (NOT the rewrite)](https://github.com/leptos-rs/leptos/pull/2158)

But I haven't been able to figure it out.

I've tried something like this:

```
#[server(StreamCounter, "/api")]
async fn counter() -> Result<impl Stream<Item = u8>, ServerFnError> {
    let counter = 
        futures::
        stream::
        futures_unordered::
        FuturesUnordered::
        from_iter(
        (0u8..10).map(
            |seconds_wait| async move {
                tokio::
                time::
                sleep(
                    Duration::from_secs(seconds_wait as u64)
                ).await;

                seconds_wait
            }
        )
    );

    Ok(counter)
}
```

which is supposed to stream an incrementing value to the client every second, but it doesn't compile because of these errors:

```
error[E0271]: expected `impl Future<Output = Result<impl Stream<Item = u8>, ServerFnError>> + Send` to be a future that resolves to `Result<<StreamCounter as ServerFn>::Output, ServerFnError>`, but it resolves to `Result<impl Stream<Item = u8>, ServerFnError>`
   --> src\app.rs:67:1
    |
67  | #[server(StreamCounter, "/api")]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Result<<... as ServerFn>::Output, ...>`, found `Result<impl Stream<Item = u8>, ...>`
68  | async fn counter() -> Result<impl Stream<Item = u8>, ServerFnError> {
    |                              ----------------------
    |                              |
    |                              the expected opaque type
    |                              the found opaque type
    |
    = note: expected enum `Result<<StreamCounter as ServerFn>::Output, _>`
               found enum `Result<impl futures::Stream<Item = u8>, _>`
    = note: distinct uses of `impl Trait` result in different opaque types
note: required by a bound in `ServerFn::{synthetic#0}`
   --> C:\Users\Elijah Ang\.cargo\registry\src\index.crates.io-6f17d22bba15001f\server_fn-0.6.15\src/lib.rs:237:22
    |
237 |     ) -> impl Future<Output = Result<Self::Output, ServerFnError<Self::Error>>> + Send;
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `ServerFn::{synthetic#0}`
    = note: this error originates in the attribute macro `server` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I'd appreciate any help. 
Thanks!

**Comment by benwis** [comment, marked as answer]
Hi eliaxelang007!
You'll want to take a look at our server fns example here:
https://github.com/leptos-rs/leptos/blob/8f8f3e23e4c4695c5053a6b0f011c90e4f64e09e/examples/server_fns_axum/src/app.rs#L487
The available codes are in the docs for the server_fn crate, I think you're looking for this one:
https://docs.rs/server_fn/latest/server_fn/codec/struct.ByteStream.html

The short of it is to set the output encoding to Streaming, and then return the ByteStream type, much like the example does with TextStream.

**Reply by eliaxelang007** [reply]
Thank you for the response! 

I didn't know that you had to set the output encoding of the server_fn to streaming, so I'm going to try that right now. 
And I think instead of returning the ByteStream type, I'm going to try the [StreamingJson](https://docs.rs/server_fn/latest/server_fn/codec/struct.StreamingJson.html) type because I ultimately want to be streaming structs of data from server to client.

I think it's starting to click into place for me.  Thanks again!

**Reply by benwis** [reply]
Here's the correct example link, my bad: https://github.com/leptos-rs/leptos/blob/8f8f3e23e4c4695c5053a6b0f011c90e4f64e09e/examples/server_fns_axum/src/app.rs#L487
